src/Compilers/Core/Portable/Syntax/AbstractSyntaxNavigator.cs

```diff
@@ -51,38 +53,95 @@ internal SyntaxToken GetNextToken(SyntaxToken current, Func<SyntaxToken, bool> p
             return GetNextToken(current, predicate, stepInto != null, stepInto);
         }
 
+        private static readonly ObjectPool<Stack<ChildSyntaxList.Enumerator>> childEnumeratorStackPool
+            = new ObjectPool<Stack<ChildSyntaxList.Enumerator>>(() => new Stack<ChildSyntaxList.Enumerator>(), 10);
+
         internal SyntaxToken GetFirstToken(SyntaxNode current, Func<SyntaxToken, bool> predicate, Func<SyntaxTrivia, bool> stepInto)
         {
-            foreach (var child in current.ChildNodesAndTokens())
+            var stack = childEnumeratorStackPool.Allocate();
+            try
             {
-                var token = child.IsToken
-                    ? GetFirstToken(child.AsToken(), predicate, stepInto)
-                    : GetFirstToken(child.AsNode(), predicate, stepInto);
+                stack.Push(current.ChildNodesAndTokens().GetEnumerator());
 
-                if (token.RawKind != None)
+                while (stack.Count > 0)
                 {
-                    return token;
+                    var en = stack.Pop();
+                    if (en.MoveNext())
+                    {
+                        var child = en.Current;
+
+                        if (child.IsToken)
+                        {
+                            var token = GetFirstToken(child.AsToken(), predicate, stepInto);
+                            if (token.RawKind != None)
+                            {
+                                return token;
+                            }
+                        }
+
+                        // push this enumerator back, not done yet
+                        stack.Push(en);
+
+                        if (child.IsNode)
+                        {
+                            stack.Push(child.AsNode().ChildNodesAndTokens().GetEnumerator());
+                        }
+                    }
                 }
-            }
 
-            return default(SyntaxToken);
+                return default(SyntaxToken);
+            }
+            finally
+            {
+                stack.Clear();
+                childEnumeratorStackPool.Free(stack);
```

Comment:
Might want to put an upper bound on the size of stack that's returned to the pool. e.g. 1024
i.e.
```C#
if (stack.Count < 1024)
{
    stack.Clear();
    childEnumeratorStackPoo.Free(stack);
}
else
{
    childEnumeratorStackPool.ForgetTrackedObject(stack);
}